Search registered S3 method for melt before redirecting to reshape2

odelmarcelle opened the pull request:
Follow-up of #3633 and #3763

The condition leading to `UseMethod` in the `melt` generic is not true for non-`data.table` object. Hence, S3 methods for `melt` defined by other packages may be skipped.

I initially suggested changing that condition in order to search if a S3 method exists for the current object before redirecting to `reshape2`.
After giving it a second thought, it appears more correct to performs the redirection to `reshape2` in a `default` method. This leaves the generic clean and allows registering new methods.

Since there were no tests on the generic `melt`, I left the test file as-is. 

After the change, using S3 method dispatch on a custom object works as intended:
```
library(data.table)
# Creating a S3 object and it's melt method
S3Object = list(dt1 = data.table(id = 1:2, a = 3L), dt2 = data.table(id = 1:2, b = 4:5))
class(S3Object) = "S3Class"
new_melt = function(data, ..., na.rm = FALSE, value.name = "value") {
  melt(Reduce(merge, data), id.vars = "id")
}
.S3method("melt", "S3Class", new_melt)

melt(S3Object) ## Working fine
```

jangorecki commented on R/fmelt.R:
```diff
@@ -3,19 +3,19 @@
+  data_name = deparse(substitute(data))
+  ns = tryCatch(getNamespace("reshape2"), error=function(e)
+    stop("The melt generic in data.table has been passed a ", class(data)[1L],", but data.table::melt currently only has a method for data.tables. Please confirm your input is a data.table, with setDT(", data_name, ") or as.data.table(", data_name, "). If you intend to use a method from reshape2, try installing that package first, but do note that reshape2 is deprecated and you should be migratin ...
+  warning("The melt generic in data.table has been passed a ", class(data)[1L], " and will attempt to redirect to the relevant reshape2 method; please note that reshape2 is deprecated, and this redirection is now deprecated as well. To continue using melt methods from reshape2 while both libraries are attached, e.g. melt.list, you can prepend the namespace like reshape2::melt(", data_name, "). In ...
```
I think we are no going into new major version so this code could be simplified for raising error, as mentioned in the warning already.

odelmarcelle replied:
Do you mean stopping the redirection towards `reshape2`?

odelmarcelle commented:
Updating this again as I still think having melt as a generic method make sense. For example, it could serve to melt an S3 object containing multiple tables and matrices into a single long "melted" format.

I have added a note regarding the change in the NEWS and I added a test on melt.default in test.Rraw.

mattdowle commented:
Thanks @odelmarcelle. Please let me know if your name is different to your id, for the contributor list displayed on CRAN. Have invited you to be project member too: it should be a button in your GitHub profile or projects page that you need to click to accept.

odelmarcelle commented:
Thank you for reviewing and merging @mattdowle. Although I'm not sure I deserve to be listed in the contributor list for such a small input, my actual name is different from my Github username.

Could you perhaps correct my name to "Olivier Delmarcelle"? I corrected the DESCRIPTION file on my forked branch, but I'm not sure how to propagate this to the master branch now that this pull request is merged. Sorry for the trouble!

mattdowle commented:
Thanks @odelmarcelle, done in 76cfe11044ed0f773b1e9f6dd4d9dfab2126f74f.
